File: src/core/runner/http-parser.ts
/**
 * Parse .http files into structured request definitions
 *
 * Format:
 *   # comments
 *   METHOD url
 *   Header: value
 *   Header2: value
 *
 *   { json body }
 */

export interface ParsedHttpRequest {
  method: string
  url: string
  headers: Record<string, string>
  body?: string
  comments: string[]
}

export function parseHttpFile(content: string): ParsedHttpRequest {
  const lines = content.split('\n')
  const comments: string[] = []
  let method = ''
  let url = ''
  const headers: Record<string, string> = {}
  let bodyLines: string[] = []
  let phase: 'comments' | 'request-line' | 'headers' | 'body' = 'comments'

  for (const line of lines) {
    const trimmed = line.trim()

    // Skip empty lines in comments/request-line phase
    if (phase === 'comments') {
      if (trimmed === '') continue
      if (trimmed.startsWith('#')) {
        comments.push(trimmed.slice(1).trim())
        continue
      }
      // First non-comment, non-empty line is the request line
      phase = 'request-line'
    }

    if (phase === 'request-line') {
      const match = trimmed.match(/^(GET|POST|PUT|DELETE|PATCH|HEAD|OPTIONS)\s+(.+)$/i)
      if (match) {
        method = match[1].toUpperCase()
        url = match[2].trim()
        phase = 'headers'
        continue
      }
      // If we don't match a request line, skip
      continue
    }

    if (phase === 'headers') {
      // Empty line separates headers from body
      if (trimmed === '') {
        phase = 'body'
        continue
      }

      // Header: value
      const headerMatch = trimmed.match(/^([\w-]+)\s*:\s*(.+)$/)
      if (headerMatch) {
        headers[headerMatch[1]] = headerMatch[2].trim()
        continue
      }

      // If it doesn't look like a header, it's probably the body start
      phase = 'body'
      bodyLines.push(line)
      continue
    }

    if (phase === 'body') {
      bodyLines.push(line)
    }
  }

  // Trim trailing empty lines from body
  while (bodyLines.length > 0 && bodyLines[bodyLines.length - 1].trim() === '') {
    bodyLines.pop()
  }

  const body = bodyLines.length > 0 ? bodyLines.join('\n') : undefined

  return { method, url, headers, body, comments }
}

File: carmelia-desktop/internal/services/httpparser.go
package services

import (
	"carmelia-desktop/internal/models"
	"regexp"
	"strings"
)

var methodRegex = regexp.MustCompile(`(?i)^(GET|POST|PUT|DELETE|PATCH|HEAD|OPTIONS)\s+(.+)$`)
var headerRegex = regexp.MustCompile(`^([\w-]+)\s*:\s*(.+)$`)

func ParseHttpFile(content string) models.ParsedHttpRequest {
	lines := strings.Split(content, "\n")
	comments := []string{}
	method := ""
	url := ""
	headers := map[string]string{}
	bodyLines := []string{}
	phase := "comments" // comments | request-line | headers | body

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)

		if phase == "comments" {
			if trimmed == "" {
				continue
			}
			if strings.HasPrefix(trimmed, "#") {
				comments = append(comments, strings.TrimSpace(trimmed[1:]))
				continue
			}
			phase = "request-line"
		}

		if phase == "request-line" {
			match := methodRegex.FindStringSubmatch(trimmed)
			if match != nil {
				method = strings.ToUpper(match[1])
				url = strings.TrimSpace(match[2])
				phase = "headers"
				continue
			}
			continue
		}

		if phase == "headers" {
			if trimmed == "" {
				phase = "body"
				continue
			}

			headerMatch := headerRegex.FindStringSubmatch(trimmed)
			if headerMatch != nil {
				headers[headerMatch[1]] = strings.TrimSpace(headerMatch[2])
				continue
			}

			// Not a header — start of body
			phase = "body"
			bodyLines = append(bodyLines, line)
			continue
		}

		if phase == "body" {
			bodyLines = append(bodyLines, line)
		}
	}

	// Trim trailing empty lines from body
	for len(bodyLines) > 0 && strings.TrimSpace(bodyLines[len(bodyLines)-1]) == "" {
		bodyLines = bodyLines[:len(bodyLines)-1]
	}

	body := ""
	if len(bodyLines) > 0 {
		body = strings.Join(bodyLines, "\n")
	}

	return models.ParsedHttpRequest{
		Method:   method,
		URL:      url,
		Headers:  headers,
		Body:     body,
		Comments: comments,
	}
}
